suprb/optimizer/solution/ga/base.py

```diff
@@ -78,30 +110,291 @@ def _optimize(self, X: np.ndarray, y: np.ndarray):
+    init: SolutionInit
+    archive: SolutionArchive
+    random_state : int, RandomState instance or None, default=None
+        Pass an int for reproducible results across multiple function calls.
+    warm_start: bool
+        If False, solutions are generated new for every `optimize()` call.
+        If True, solutions are used from previous runs.
+    n_jobs: int
+        The number of threads / processes the optimization uses.
+    """
+
+    n_elitists_: int
+
+    def __init__(
+        self,
+        n_iter: int = 32,
+        population_size: int = 32,
+        elitist_ratio: float = 0.17,
+        mutation: SolutionMutation = BitFlips(),
+        crossover: SolutionCrossover = NPoint(n=3),
+        selection: SolutionSelection = Tournament(),
+        v_min: float = 0.005,
+        v_max: float = 0.15,
+        mutation_rate: float = 0.025,
+        mutation_rate_min: float = 0.001,
+        mutation_rate_max: float = 0.25,
+        mutation_rate_multiplier=1.1,
+        crossover_rate: float = 0.75,
+        crossover_rate_min: float = 0.5,
+        crossover_rate_max: float = 1.0,
+        crossover_rate_multiplier=1.1,
+        init: SolutionInit = RandomInit(),
+        archive: SolutionArchive = Elitist(),
+        random_state: int = None,
+        n_jobs: int = 1,
+        warm_start: bool = True,
+    ):
+        super().__init__(
+            n_iter=n_iter,
+            population_size=population_size,
+            init=init,
+            archive=archive,
+            random_state=random_state,
+            n_jobs=n_jobs,
+            warm_start=warm_start,
+        )
+
+        self.v_min = v_min
+        self.v_max = v_max
+        self.mutation_rate = mutation_rate
+        self.mutation_rate_min = mutation_rate_min
+        self.mutation_rate_max = mutation_rate_max
+        self.mutation_rate_multiplier = mutation_rate_multiplier
+        self.crossover_rate = crossover_rate
+        self.crossover_rate_min = crossover_rate_min
+        self.crossover_rate_max = crossover_rate_max
+        self.crossover_rate_multiplier = crossover_rate_multiplier
+        self.mutation = mutation
+        self.crossover = crossover
+        self.selection = selection
+        self.elitist_ratio = elitist_ratio
+
+    def adjust_rates(self):
+        gdm = np.mean([i.fitness_ for i in self.population_]) / np.max([i.fitness_ for i in self.population_])
+        if gdm > self.v_max:
+            self.mutation_rate = min(self.mutation_rate_max, self.mutation_rate * self.mutation_rate_multiplier)
+            self.crossover_rate = max(self.crossover_rate_min, self.crossover_rate / self.crossover_rate_multiplier)
+        elif gdm < self.v_min:
+            self.mutation_rate = max(self.mutation_rate_min, self.mutation_rate / self.mutation_rate_multiplier)
+            self.crossover_rate = min(self.crossover_rate_max, self.crossover_rate * self.crossover_rate_multiplier)
+
+
+class SelfAdaptingGeneticAlgorithm2(GeneticAlgorithm):
```

Comment:
why is this here while saga3 is in the other folder?

Reply:
For SAGA1 and SAGA2 it was possible to generalize enough to use the functions from GA, but for SAGA3 and SAS not. I will move all the SAGA into the same directory (outside GA)